server/src/services/documentSymbol/onDocumentSymbol.ts

```diff
@@ -71,50 +71,91 @@ export function onDocumentSymbol(serverState: ServerState) {
-          new InterfaceDefinition(document, builder),
-          new FunctionDefinition(document, builder),
-          new ContractDefinition(document, builder),
-          new EventDefinition(document, builder),
-          new StateVariableDeclaration(document, builder),
-          new VariableDeclaration(document, builder),
-          new ConstantDefinition(document, builder),
-          new ConstructorDefinition(document, builder),
-          new EnumDefinition(document, builder),
-          new ErrorDefinition(document, builder),
-          new FallbackFunctionDefinition(document, builder),
-          new LibraryDefinition(document, builder),
-          new ModifierDefinition(document, builder),
-          new ReceiveFunctionDefinition(document, builder),
-          new UserDefinedValueTypeDefinition(document, builder),
-          new YulFunctionDefinition(document, builder),
-          new UnnamedFunctionDefinition(document, builder),
+        const finders: SymbolFinder[] = [
+          new StructDefinition(),
+          new StructMember(),
+          new InterfaceDefinition(),
+          new FunctionDefinition(),
+          new ContractDefinition(),
+          new EventDefinition(),
+          new StateVariableDefinition(),
+          new VariableDeclarationStatement(),
+          new ConstantDefinition(),
+          new ConstructorDefinition(),
+          new EnumDefinition(),
+          new ErrorDefinition(),
+          new FallbackFunctionDefinition(),
+          new LibraryDefinition(),
+          new ModifierDefinition(),
+          new ReceiveFunctionDefinition(),
+          new UserDefinedValueTypeDefinition(),
+          new YulFunctionDefinition(),
+          new UnnamedFunctionDefinition(),
         ];
 
-        const indexedVisitors = _.keyBy(visitors, "ruleKind");
-
         const cursor = parseOutput.createTreeCursor();
-        const ruleKinds = visitors.map((v) => v.ruleKind);
 
         // Useful to keep this here for development
-        // const kursor: Cursor = parseTree.cursor.clone();
+        // const kursor = cursor.clone();
         // do {
         //   console.log(
-        //     `${"  ".repeat(kursor.pathRuleNodes.length)}${kursor.node.kind}(${
-        //       ["R", "T"][kursor.node.type]
-        //     }): ${kursor.node?.text ?? ""}`
+        //     `${"  ".repeat(kursor.ancestors().length)}${kursor.node().kind}(${
+        //       // ["R", "T"][kursor.node().type]
+        //       kursor.node().type
+        //     })(${kursor.label}): ${(kursor.node() as TokenNode)?.text || ""}`
         //   );
         // } while (kursor.goToNext());
 
-        span = transaction.startChild({ op: "walk-generate-symbols" });
-        while (cursor.goToNextRuleWithKinds(ruleKinds)) {
-          const node = cursor.node() as RuleNode;
+        span = transaction.startChild({ op: "run-query" });
+
+        // Execute a single call with all the queries
+        const queries = finders.map((v) => Query.parse(v.query));
```

Comment:
At the moment, if any query parsing fails, our protocol tests fails as well. Even thought the full error description is not shown on the test run, it's logged when running the LSP. Should I write unit tests for each query ? I'm wondering if the value is worth the effort

Reply:
I'm only suggesting we add a unit test for ALL queries in the feature (so two tests for two features), so that we get the error in the test failure message, rather than hidden in LSP traces in runtime.
Basically something like this:

```ts
// refactor this list here into a separate function
export function createFinders(): SymbolFinder[] {
  return [
          new StructDefinition(),
          new StructMember(),
          // ...etc
        ];
}

// then later just call it from a single unit test
it("can create queries correctly", () => {
  for (const finders of createFinders()) {
    // assert that parsing the query didn't throw.
  }
});
```

Reply:
I'm happy to address this later. No need to block this. Thanks.